execute_feasibility_test.py

```diff
@@ -0,0 +1,318 @@
+
+    if column_attribute == "module":
+        module_names = list(filter(lambda x: x != "percentage_fulfilled", column_names))
+        append_percentage_fulfilled_column(site_module_yes_no_map, module_names, optional_modules)
+    return site_module_yes_no_map
+
+
+def get_column_keys(report_results, column_attribute):
+    column_keys = [report[column_attribute] for report in report_results]
+
+    if column_attribute == "module":
+        optional_modules = [query_report["module"] for query_report in filter(lambda x: x["optional"] is True, report_results)]
+        sorted_query_modules = []
+        for module in column_keys:
+            if module not in optional_modules:
+                sorted_query_modules.append(module)
+
+        sorted_query_modules += optional_modules
+        sorted_query_modules += ["percentage_fulfilled"]
+
+        return sorted_query_modules, optional_modules
+    else:
+        return column_keys, []
+
+
+def beautyfy_result(result):
+    if result == "yes":
+        return "Yes"
+    elif result == "no":
+        return "No"
+    else:
+        return result
+
+
+# column_attribute can be date or module
+def convert_to_table(report_results, column_attribute):
+    row_keys = ["total_yes", "total_no", "total_na"] + list(DSF_SITE_IDS)
+    column_keys, optional_column_keys = get_column_keys(report_results, column_attribute)
+    columns = create_columns(report_results, column_keys, optional_column_keys, column_attribute)
+
+    pTable = PrettyTable()
+    first_column = ["Total Yes | 100% fullfilled", "Total No | >0% fullfilled", "Total Na | 0% fullfilled"] + list(DSF_SITE_IDS)
+    pTable.add_column("Site URL", first_column)
+
+    for column_key in column_keys:
+        column = []
+        for row_key in row_keys:
+            column.append(beautyfy_result(columns[column_key][row_key]))
+
+        column_name = column_key + " (optional)" if column_key in optional_column_keys else column_key
+        if column_name == "percentage_fulfilled":
+            column_name = "% fulfilled"
+        pTable.add_column(column_name, column)
+
+    return pTable
+
+
+def load_history_report():
+    history_query = load_json_file(HISTORY_QUERY_FILE)
+
+    if os.path.isfile(HISTORY_REPORT_FILE):
+        with open(HISTORY_REPORT_FILE, "r") as f:
+            history_report = json.load(f)
+
+            if history_report["query"] != history_query:
+                raise ValueError("history report contains query that is not equal to the current history query")
+
+            return history_report
+    else:
+        open(HISTORY_REPORT_FILE, "x")
+        history_report = {"reports": [], "query": history_query}
+        return history_report
+
+
+def update_and_save_history_report(history_report, new_result):
+    now = datetime.now().strftime("%Y-%m-%d")
+    new_history_result = {"date": now}
+    new_history_result["result"] = new_result
+
+    is_new_date = True
+    for i in range(len(history_report["reports"])):
```

Comment:
@bastianschaffer  this seems very complicated - can you not simply check the last entry in the array for the date?

Reply:
Currently I assume that the element with the most recent date could be anywhere in the list.
Is it save to assume that the order in which the elements are stored will always be preserved by the serializers?